### DISCUSSION TITLE
Namespace and Class Directives

### DISCUSSION THREAD

**Discussion by TylerDM** [dotnet/csharplang#4177] [Language Ideas]
# Namespace and Class Directives

## Summary
Allow `namespace` and `class` to have Using Directive-like syntax.

## Motivation
- Primary
  - Remove boilerplate braces found in 99% of C# files.
  - Remove boilerplate indentation found in 99% of C# files.
  - Removing 4-5 lines of code means that we see more of the file immediately upon opening.
- Secondary
  - Reduce line count at no real world cost.  It's unfortunate that LoC is used as a decision making metric, but it does happen.

## Detailed design
If you've ever taken a C# course, you've probably seen a class like this at least once.
```csharp
using System;

namespace Animals
{
    public class Dog
    {
        private readonly string _breed;

        public Dog(string breed)
        {
            _breed = breed;
        }

        public void Bark() => Console.WriteLine($"The {breed} dog barks.");
    }    
}
```

I propose removal of the braces from `using` and `class` and assume they contain the entire file;
```csharp
using System;

namespace Animals;
public class Dog;

private readonly string _breed;

public Dog(string breed)
{
    _breed = breed;
}

public void Bark() => Console.WriteLine($"The {breed} dog barks.");
```

This is quite a lot better!  We've eliminated the boiler plate code that never changes in 99% of C# files.  We've also eliminated 2 tabs on nearly the entire file.  Costs?  None.  We can take this just a hair further as well:

The basic concept works well for `record` and other similar concepts.
```csharp
using System;

public record Animals.Dog(string Breed);
```

## Drawbacks
I personally do not like having multiple syntaxes for the same feature.  In this case, however, I think it makes sense.  You can use the old syntax for multiple namespaces or classes in a single file, should you want to.  The old syntax would also be useful for nested classes.  And then use the new syntax for the 99% of the time that those things aren't desired or needed.

This does reinforce a specific coding style of having one class per file.  I have mixed feelings about catering C# to any particular style or use case.  That said, the one class per file style is so ubiquitous that I believe the pragmatic benefits outweigh the philosophical drawbacks.

## Alternatives
Combine the `namespace` and `class` directives into a single directive.
```csharp
using System;

public class Animals.Dog;

private readonly string _breed;

public void Bark() => Console.WriteLine($"The {breed} dog barks.");

public Dog(string breed)
{
    _breed = breed;
}
```

You could also allow multiple namespaces and class per file by allowing multiple directives, with the latest one taking precedence.
```csharp
using System;

namespace Animals;

public abstract class Pet;
public abstract void Vocalize();

public class Dog : Pet;
public override void Vocalize() => Console.WriteLine("Woof!");

public class Cat : Pet;
public override void Vocalize() => Console.WriteLine("Meow!");
```

## Unresolved questions
None known.

## Design meetings
None thus far.

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/137

I don't see that being extended to `class`.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/4141
